File: server/chat/conversation-manager.js
/* =========================================
    会話履歴・コンテキスト管理
   ========================================= */

/*
## 概要
会話履歴の管理、コンテキスト情報の整理、セッション状態の追跡を行う責任を持つ。

## 責任
- 会話履歴の管理と最適化
- コンテキスト情報の整理と拡張
- セッション状態の追跡
- メモリ使用量の最適化
*/

export class ConversationManager {
    constructor() {
        this.maxHistoryItems = 15;
        this.maxHistoryLength = 10000; // 文字数制限
        this.contextDefaults = {
            currentPath: '/workspace',
            fileList: [],
            currentFile: null,
            openFileInfo: null
        };
    }

    /**
     * コンテキストを準備・拡張
     */
    async prepareContext(context = {}) {
        // デフォルト値との統合
        const enrichedContext = {
            ...this.contextDefaults,
            ...context
        };

        // 会話履歴の最適化
        enrichedContext.conversationHistory = this._optimizeConversationHistory(
            enrichedContext.conversationHistory || []
        );

        // ファイル一覧の整理
        enrichedContext.fileList = this._normalizeFileList(enrichedContext.fileList);

        // カスタムプロンプトの状態確認
        enrichedContext.customPrompt = this._validateCustomPrompt(enrichedContext.customPrompt);

        // コンテキストのメタデータを追加
        enrichedContext.metadata = this._generateContextMetadata(enrichedContext);

        return enrichedContext;
    }

    /**
     * 新しいチャットを提案すべきかどうかを判定
     */
    shouldSuggestNewChat(context) {
        const history = context.conversationHistory || [];
        
        // 履歴数ベースの判定
        if (history.length >= this.maxHistoryItems) {
            return true;
        }

        // 文字数ベースの判定
        const totalLength = this._calculateTotalHistoryLength(history);
        if (totalLength >= this.maxHistoryLength) {
            return true;
        }

        // コンテキストスイッチの検出
        if (this._detectContextSwitch(history)) {
            return true;
        }

        return false;
    }

    /**
     * 会話履歴の最適化
     */
    _optimizeConversationHistory(history) {
        if (!Array.isArray(history)) {
            return [];
        }

        // 履歴数の制限
        let optimizedHistory = history.slice(-this.maxHistoryItems);

        // 文字数制限の適用
        const totalLength = this._calculateTotalHistoryLength(optimizedHistory);
        if (totalLength > this.maxHistoryLength) {
            optimizedHistory = this._truncateHistoryByLength(optimizedHistory);
        }

        // 重複の除去
        optimizedHistory = this._removeDuplicateHistory(optimizedHistory);

        // 不完全な履歴エントリの除去
        optimizedHistory = this._removeIncompleteHistory(optimizedHistory);

        return optimizedHistory;
    }

    /**
     * ファイル一覧の正規化
     */
    _normalizeFileList(fileList) {
        if (!Array.isArray(fileList)) {
            return [];
        }

        return fileList
            .filter(file => file && typeof file === 'string')
            .map(file => file.trim())
            .filter(file => file.length > 0)
            .slice(0, 1000); // 上限設定
    }

    /**
     * カスタムプロンプトの検証
     */
    _validateCustomPrompt(customPrompt) {
        if (!customPrompt || typeof customPrompt !== 'object') {
            return null;
        }

        return {
            enabled: !!customPrompt.enabled,
            name: customPrompt.name || 'Unknown',
            content: customPrompt.content || '',
            description: customPrompt.description || ''
        };
    }

    /**
     * コンテキストメタデータの生成
     */
    _generateContextMetadata(context) {
        return {
            timestamp: new Date().toISOString(),
            historyCount: (context.conversationHistory || []).length,
            fileCount: (context.fileList || []).length,
            hasCustomPrompt: !!(context.customPrompt?.enabled),
            hasOpenFile: !!context.currentFile,
            contextSize: this._calculateContextSize(context)
        };
    }

    /**
     * 会話履歴の総文字数を計算
     */
    _calculateTotalHistoryLength(history) {
        return history.reduce((total, entry) => {
            const userLength = entry.user ? entry.user.length : 0;
            const aiLength = entry.ai ? entry.ai.length : 0;
            return total + userLength + aiLength;
        }, 0);
    }

    /**
     * 文字数制限による履歴の切り詰め
     */
    _truncateHistoryByLength(history) {
        const truncated = [];
        let currentLength = 0;

        // 新しい履歴から順に追加
        for (let i = history.length - 1; i >= 0; i--) {
            const entry = history[i];
            const entryLength = (entry.user?.length || 0) + (entry.ai?.length || 0);
            
            if (currentLength + entryLength <= this.maxHistoryLength) {
                truncated.unshift(entry);
                currentLength += entryLength;
            } else {
                break;
            }
        }

        return truncated;
    }

    /**
     * 重複履歴の除去
     */
    _removeDuplicateHistory(history) {
        const seen = new Set();
        return history.filter(entry => {
            const key = `${entry.user || ''}_${entry.ai || ''}`;
            if (seen.has(key)) {
                return false;
            }
            seen.add(key);
            return true;
        });
    }

    /**
     * 不完全な履歴エントリの除去
     */
    _removeIncompleteHistory(history) {
        return history.filter(entry => {
            // ユーザーメッセージは必須
            if (!entry.user || entry.user.trim().length === 0) {
                return false;
            }
            
            // AI応答が極端に短い場合は除去
            if (entry.ai && entry.ai.trim().length < 3) {
                return false;
            }

            return true;
        });
    }

    /**
     * コンテキストスイッチの検出
     */
    _detectContextSwitch(history) {
        if (history.length < 3) {
            return false;
        }

        // 最近の会話のトピックを分析
        const recentMessages = history.slice(-3).map(entry => entry.user);
        
        // 簡単なキーワードベースの分析
        const keywords = [
            '新しい', '別の', '違う', 'change', 'switch', '切り替え', 
            'help', 'ヘルプ', '使い方', '機能'
        ];

        const lastMessage = recentMessages[recentMessages.length - 1].toLowerCase();
        return keywords.some(keyword => lastMessage.includes(keyword));
    }

    /**
     * コンテキストサイズの計算（概算）
     */
    _calculateContextSize(context) {
        let size = 0;
        
        // 基本フィールドのサイズ
        size += (context.currentPath || '').length;
        size += JSON.stringify(context.fileList || []).length;
        size += (context.currentFile || '').length;
        size += (context.openFileInfo || '').length;
        
        // 会話履歴のサイズ
        size += this._calculateTotalHistoryLength(context.conversationHistory || []);
        
        // カスタムプロンプトのサイズ
        if (context.customPrompt) {
            size += (context.customPrompt.content || '').length;
            size += (context.customPrompt.name || '').length;
            size += (context.customPrompt.description || '').length;
        }

        return size;
    }

    /**
     * 履歴に新しいエントリを追加
     */
    addHistoryEntry(context, userMessage, aiResponse) {
        const history = context.conversationHistory || [];
        
        const newEntry = {
            user: userMessage,
            ai: aiResponse,
            timestamp: new Date().toISOString()
        };

        history.push(newEntry);
        context.conversationHistory = this._optimizeConversationHistory(history);
        
        return context;
    }

    /**
     * 履歴をクリア
     */
    clearHistory(context) {
        context.conversationHistory = [];
        return context;
    }

    /**
     * 統計情報の取得
     */
    getHistoryStats(context) {
        const history = context.conversationHistory || [];
        
        return {
            totalEntries: history.length,
            totalLength: this._calculateTotalHistoryLength(history),
            shouldSuggestNewChat: this.shouldSuggestNewChat(context),
            oldestEntry: history.length > 0 ? history[0].timestamp : null,
            newestEntry: history.length > 0 ? history[history.length - 1].timestamp : null
        };
    }

    /**
     * 設定の取得/更新
     */
    getConfiguration() {
        return {
            maxHistoryItems: this.maxHistoryItems,
            maxHistoryLength: this.maxHistoryLength,
            contextDefaults: { ...this.contextDefaults }
        };
    }

    updateConfiguration(config) {
        if (config.maxHistoryItems && config.maxHistoryItems > 0) {
            this.maxHistoryItems = Math.min(config.maxHistoryItems, 50); // 上限設定
        }
        
        if (config.maxHistoryLength && config.maxHistoryLength > 0) {
            this.maxHistoryLength = Math.min(config.maxHistoryLength, 50000); // 上限設定
        }

        if (config.contextDefaults && typeof config.contextDefaults === 'object') {
            this.contextDefaults = { ...this.contextDefaults, ...config.contextDefaults };
        }
    }

    /**
     * マネージャーの状態取得
     */
    getStatus() {
        return {
            maxHistoryItems: this.maxHistoryItems,
            maxHistoryLength: this.maxHistoryLength,
            contextDefaults: this.contextDefaults,
            isHealthy: true,
            lastActivity: new Date().toISOString()
        };
    }
}

File: server_pythonver/chat/conversation_manager.py
import datetime
import json

class ConversationManager:
    def __init__(self):
        self.max_history_items = 15
        self.max_history_length = 10000  # 文字数制限
        self.context_defaults = {
            "currentPath": "/workspace",
            "fileList": [],
            "currentFile": None,
            "openFileInfo": None
        }

    async def prepare_context(self, context: dict = None):
        if context is None:
            context = {}

        enriched_context = {
            **self.context_defaults,
            **context
        }

        enriched_context["conversationHistory"] = self._optimize_conversation_history(
            enriched_context.get("conversationHistory", [])
        )

        enriched_context["fileList"] = self._normalize_file_list(enriched_context.get("fileList", []))

        enriched_context["customPrompt"] = self._validate_custom_prompt(enriched_context.get("customPrompt"))

        enriched_context["metadata"] = self._generate_context_metadata(enriched_context)

        return enriched_context

    def should_suggest_new_chat(self, context: dict):
        history = context.get("conversationHistory", [])

        if len(history) >= self.max_history_items:
            return True

        total_length = self._calculate_total_history_length(history)
        if total_length >= self.max_history_length:
            return True

        if self._detect_context_switch(history):
            return True

        return False

    def _optimize_conversation_history(self, history: list):
        if not isinstance(history, list):
            return []

        optimized_history = history[-self.max_history_items:]

        total_length = self._calculate_total_history_length(optimized_history)
        if total_length > self.max_history_length:
            optimized_history = self._truncate_history_by_length(optimized_history)

        optimized_history = self._remove_duplicate_history(optimized_history)

        optimized_history = self._remove_incomplete_history(optimized_history)

        return optimized_history

    def _normalize_file_list(self, file_list: list):
        if not isinstance(file_list, list):
            return []

        return [
            file.strip() for file in file_list
            if isinstance(file, str) and len(file.strip()) > 0
        ][:1000]

    def _validate_custom_prompt(self, custom_prompt: dict):
        if not isinstance(custom_prompt, dict):
            return None

        return {
            "enabled": bool(custom_prompt.get("enabled")),
            "name": custom_prompt.get("name", "Unknown"),
            "content": custom_prompt.get("content", ""),
            "description": custom_prompt.get("description", "")
        }

    def _generate_context_metadata(self, context: dict):
        return {
            "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
            "historyCount": len(context.get("conversationHistory", [])),
            "fileCount": len(context.get("fileList", [])),
            "hasCustomPrompt": bool(context.get("customPrompt", {}).get("enabled")),
            "hasOpenFile": bool(context.get("currentFile")),
            "contextSize": self._calculate_context_size(context)
        }

    def _calculate_total_history_length(self, history: list):
        total = 0
        for entry in history:
            user_length = len(entry.get("user", ""))
            ai_length = len(entry.get("ai", ""))
            total += user_length + ai_length
        return total

    def _truncate_history_by_length(self, history: list):
        truncated = []
        current_length = 0

        for entry in reversed(history):
            entry_length = len(entry.get("user", "")) + len(entry.get("ai", ""))
            if current_length + entry_length <= self.max_history_length:
                truncated.insert(0, entry)
                current_length += entry_length
            else:
                break
        return truncated

    def _remove_duplicate_history(self, history: list):
        seen = set()
        unique_history = []
        for entry in history:
            key = f"{entry.get('user', '')}_{entry.get('ai', '')}"
            if key not in seen:
                seen.add(key)
                unique_history.append(entry)
        return unique_history

    def _remove_incomplete_history(self, history: list):
        return [
            entry for entry in history
            if entry.get("user") and len(entry["user"].strip()) > 0 and \
               (not entry.get("ai") or len(entry["ai"].strip()) >= 3)
        ]

    def _detect_context_switch(self, history: list):
        if len(history) < 3:
            return False

        recent_messages = [entry.get("user", "").lower() for entry in history[-3:]]

        keywords = [
            "新しい", "別の", "違う", "change", "switch", "切り替え",
            "help", "ヘルプ", "使い方", "機能"
        ]

        last_message = recent_messages[-1]
        return any(keyword in last_message for keyword in keywords)

    def _calculate_context_size(self, context: dict):
        size = 0

        size += len(context.get("currentPath", ""))
        size += len(json.dumps(context.get("fileList", [])))
        size += len(context.get("currentFile", ""))
        size += len(context.get("openFileInfo", ""))

        size += self._calculate_total_history_length(context.get("conversationHistory", []))

        if context.get("customPrompt"):
            size += len(context["customPrompt"].get("content", ""))
            size += len(context["customPrompt"].get("name", ""))
            size += len(context["customPrompt"].get("description", ""))

        return size

    def add_history_entry(self, context: dict, user_message: str, ai_response: str):
        history = context.get("conversationHistory", [])

        new_entry = {
            "user": user_message,
            "ai": ai_response,
            "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat()
        }

        history.append(new_entry)
        context["conversationHistory"] = self._optimize_conversation_history(history)

        return context

    def clear_history(self, context: dict):
        context["conversationHistory"] = []
        return context

    def get_history_stats(self, context: dict):
        history = context.get("conversationHistory", [])

        return {
            "totalEntries": len(history),
            "totalLength": self._calculate_total_history_length(history),
            "shouldSuggestNewChat": self.should_suggest_new_chat(context),
            "oldestEntry": history[0]["timestamp"] if history else None,
            "newestEntry": history[-1]["timestamp"] if history else None
        }

    def get_configuration(self):
        return {
            "maxHistoryItems": self.max_history_items,
            "maxHistoryLength": self.max_history_length,
            "contextDefaults": {**self.context_defaults}
        }

    def update_configuration(self, config: dict):
        if config.get("maxHistoryItems") and config["maxHistoryItems"] > 0:
            self.max_history_items = min(config["maxHistoryItems"], 50)

        if config.get("maxHistoryLength") and config["maxHistoryLength"] > 0:
            self.max_history_length = min(config["maxHistoryLength"], 50000)

        if config.get("contextDefaults") and isinstance(config["contextDefaults"], dict):
            self.context_defaults = {**self.context_defaults, **config["contextDefaults"]}

    def get_status(self):
        return {
            "maxHistoryItems": self.max_history_items,
            "maxHistoryLength": self.max_history_length,
            "contextDefaults": self.context_defaults,
            "isHealthy": True,
            "lastActivity": datetime.datetime.now(datetime.timezone.utc).isoformat()
        }
